### PR TITLE
[wxwidgets] Fix linux build

### LINKED ISSUE #23758: [wxwidgets] build failure on linux
**Host Environment**
- OS: Linux

**To Reproduce**
`./vcpkg install wxwidgets`

**Failure logs**
-Cut and paste the appropriate build messages from the console output.
-Please attach any additional failure logs mentioned in the console output.

```bash
tal@tal-pav15:~/vcpkg$ ./vcpkg install wxwidgets
Computing installation plan...
The following packages will be built and installed:
    wxwidgets[core]:x64-linux -> 3.1.5#7
Detecting compiler hash for triplet x64-linux...
Restored 0 packages from /home/tal/.cache/vcpkg/archives in 10.06 ms. Use --debug to see more details.
Starting package 1/1: wxwidgets:x64-linux
Building package wxwidgets[core]:x64-linux...
-- Using cached wxWidgets-wxWidgets-9c0a8be1dc32063d91ed1901fd5fcd54f4f955a1.tar.gz.
-- Extracting source /home/tal/vcpkg/downloads/wxWidgets-wxWidgets-9c0a8be1dc32063d91ed1901fd5fcd54f4f955a1.tar.gz
-- Applying patch disable-platform-lib-dir.patch
-- Applying patch fix-build.patch
-- Using source at /home/tal/vcpkg/buildtrees/wxwidgets/src/54f4f955a1-0d7c66a59e.clean
-- Configuring x64-linux-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:128 (message):
    Command failed: /home/tal/vcpkg/downloads/tools/cmake-3.22.2-linux/cmake-3.22.2-linux-x86_64/bin/cmake /home/tal/vcpkg/buildtrees/wxwidgets/src/54f4f955a1-0d7c66a59e.clean -DwxUSE_REGEX=builtin -DwxUSE_ZLIB=sys -DwxUSE_EXPAT=sys -DwxUSE_LIBJPEG=sys -DwxUSE_LIBPNG=sys -DwxUSE_LIBTIFF=sys -DwxBUILD_DISABLE_PLATFORM_LIB_DIR=ON -DwxUSE_STL=OFF -DwxUSE_STD_CONTAINERS=OFF -DCMAKE_MAKE_PROGRAM=/home/tal/vcpkg/downloads/tools/ninja-1.10.2-linux/ninja -DCMAKE_SYSTEM_NAME=Linux -DBUILD_SHARED_LIBS=OFF -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=/home/tal/vcpkg/scripts/toolchains/linux.cmake -DVCPKG_TARGET_TRIPLET=x64-linux -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=external -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=/home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=/home/tal/vcpkg -D_VCPKG_INSTALLED_DIR=/home/tal/vcpkg/installed -DVCPKG_MANIFEST_INSTALL=OFF -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/home/tal/vcpkg/packages/wxwidgets_x64-linux/debug
    Working Directory: /home/tal/vcpkg/buildtrees/wxwidgets/x64-linux-dbg
    Error code: 1
    See logs for more information:
      /home/tal/vcpkg/buildtrees/wxwidgets/config-x64-linux-dbg-out.log
      /home/tal/vcpkg/buildtrees/wxwidgets/config-x64-linux-dbg-err.log

Call Stack (most recent call first):
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:393 (vcpkg_execute_required_process)
  ports/wxwidgets/portfile.cmake:35 (vcpkg_cmake_configure)
  scripts/ports.cmake:145 (include)
```

[wxwidgets.zip](https://github.com/microsoft/vcpkg/files/8346940/wxwidgets.zip)

**Additional context**
Add any other context about the problem here, such as what you have already tried to resolve the issue.

I try to talk with @JackBoosY many times to solve this issue.

### PR DESCRIPTION
**Pull request by JackBoosY**
1. Fix configure issue in Linux:
```
CMake Error at build/cmake/modules/cotire.cmake:2374 (file):
  Error evaluating generator expression:

    $<COMPILE_LANG_AND_ID:CUDA,NVIDIA>

  $<COMPILE_LANG_AND_ID:lang,id> may only be used with binary targets to
  specify include directories, compile definitions, and compile options.  It
  may not be used with the add_custom_command, add_custom_target, or
  file(GENERATE) commands.
Call Stack (most recent call first):
  build/cmake/modules/cotire.cmake:2996 (cotire_generate_target_script)
  build/cmake/modules/cotire.cmake:3387 (cotire_process_target_language)
  build/cmake/modules/cotire.cmake:3571 (cotire_target)
  build/cmake/functions.cmake:391 (cotire)
  build/cmake/functions.cmake:407 (wx_target_enable_precomp)
  build/cmake/lib/net/CMakeLists.txt:39 (wx_finalize_lib)
```

Related upstream commit: https://github.com/wxWidgets/wxWidgets/commit/2f6dd5e80c3639a6702b84568b84aaf6f7473a9e

2. Fix `--libs` output:
```
-L/home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib -pthread /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_gtk3u_xrc-3.1.a /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_gtk3u_qa-3.1.a /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_baseu_net-3.1.a /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_gtk3u_html-3.1.a /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_gtk3u_core-3.1.a /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_baseu_xml-3.1.a /home/usr/work/17111/vcpkg/packages/wxwidgets_x64-linux/lib/wx/config/../../../lib/libwx_baseu-3.1.a gtk-3 gdk-3 pangocairo-1.0 pango-1.0 atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so -lwx_gtk3u_core-3.1 gtk-3 gdk-3 pangocairo-1.0 pango-1.0 atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so libjpeg.a libpng.a> libpng16d.a> libz.a m libtiff.a> libtiffd.a> liblzma.a> liblzmad.a> libjpeg.a libz.a m gtk-3 gdk-3 pangocairo-1.0 pango-1.0 atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so -lwx_baseu-3.1 libexpat.a> libexpat.a> gtk-3 gdk-3 pangocairo-1.0 pango-1.0 atk-1.0 cairo-gobject cairo gdk_pixbuf-2.0 gio-2.0 gobject-2.0 glib-2.0 libSM.so libICE.so libX11.so libXext.so libXtst.so libz.a -lwxregexu-3.1 secret-1 gio-2.0 gobject-2.0 glib-2.0 libc.so -pthread dl
```

Related upstream commit: https://github.com/wxWidgets/wxWidgets/commit/c284c88db1ec8db78fcc342641399938dc5676f5

Fixes #23758.

Note: although `wxwidgets:x64-linux` still set to `fail`, but the user can install gtk3 with other package manager, so I still prefer to fix the upstream bugs in this PR.

### CODE REVIEW

**Comment by JackBoosY** [comment]
@dg0yt I think I can now accept using other package managers with gtk3 .
After #22201 is merged, I will make gtk switch to this port.

Do you agree with this?

**Comment by JackBoosY** [comment]
cc @talregev for test this PR.

**Comment by dg0yt** [comment]
> @dg0yt I think I can now accept using other package managers with gtk3 . After #22201 is merged, I will make gtk switch to this port.
> 
> Do you agree with this?

I won't block it.

However, testing locally on Ubuntu 18.04 fails immediately, in locating the system's GTK3 development package.
What has changed since the last tests is that CMake 3.22 picks `pkgconf` from vcpkg-pkgconf.
If I replace `installed/x64-linux/tools/pkgconf/pkgconf` with a symlink to `/usr/bin/pkg-config`, configuration and build succeeds.
I assume that the configuration of the pc file search path is still not sufficiently including system directories.
Related: #23010, #23322. CC @Neumann-A 
And also #23429.

**Comment by talregev** [comment]
Can you update in the top that you waiting for #22201 as a task?
Like this:
* [ ] https://github.com/microsoft/vcpkg/pull/22201

**Comment by dg0yt** [comment]
https://github.com/JackBoosY/vcpkg/pull/6 is a workaround for https://github.com/microsoft/vcpkg/pull/23765#issuecomment-1078775013.

As usual, mixing system and vcpkg libraries could be dangerous. The `gtk+-3.0.pc` file will pull in stuff from the system which is also in vcpkg, such as include dirs:
`-- Found GTK3: /usr/include/gtk-3.0;/usr/include/at-spi2-atk/2.0;/usr/include/at-spi-2.0;/usr/include/dbus-1.0;/usr/lib/x86_64-linux-gnu/dbus-1.0/include;/usr/include/gtk-3.0;/usr/include/gio-unix-2.0/;/usr/include/cairo;/usr/include/pango-1.0;/usr/include/harfbuzz;/usr/include/pango-1.0;/usr/include/atk-1.0;/usr/include/cairo;/usr/include/pixman-1;/usr/include/freetype2;/usr/include/libpng16;/usr/include/freetype2;/usr/include/libpng16;/usr/include/gdk-pixbuf-2.0;/usr/include/libpng16;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include`

File: scripts/ci.baseline.txt [hunk 1]
```diff
@@ -1359,7 +1359,6 @@ winreg:x64-osx=fail
 wpilib:arm64-windows=fail
 wpilib:x64-osx=fail
 wxchartdir:x64-osx=fail
-wxwidgets:x64-linux=fail
```
Comment (dg0yt) [inline scripts/ci.baseline.txt:1362]: Of course, this cannot be removed until GTK3 is in vcpkg.
(And we don't want to have the system packages. The system packages are built against different dependencies.)

**Comment by talregev** [comment]
Can you rebase with master?
I need this PR that was merge:
https://github.com/microsoft/vcpkg-tool/pull/434

**Comment by JackBoosY** [comment]
@talregev Yeah, I also need consider https://github.com/JackBoosY/vcpkg/pull/3/files. Will finish this PR today.

**Comment by JackBoosY** [comment]
I decide to only fix the wxwidgets internal code in this PR.

**Comment by talregev** [comment]
@JackBoosY Thank you for merge the master inside.
I see you roll back some of the changes.
There is issues that need to be fix:
- tools/wxwidgets/wx-config is link to packages instead to lib
- tools/wxwidgets/wxrc is link to packages instead to lib
- lib/wx/config/gtk3-unicode-static-3.1 the cppflags/cflags/cxxflags option (the include folders) are linked to packages instead of include folder. 

Can you fix it?

**Comment by JackBoosY** [comment]
@talregev I'd like to fix that in another PR.

**Review by LilyWangLL — APPROVED** [review]

**Comment by talregev** [comment]
@JackBoosY 
> @talregev I'd like to fix that in another PR.

Can you link to the PR?

**Comment by talregev** [comment]
> @talregev I'd like to fix that in another PR.

@JackBoosY 
Can you make a PR with these fixes?

**Comment by talregev** [comment]
> @JackBoosY Thank you for merge the master inside. I see you roll back some of the changes. There is issues that need to be fix:
> 
> * tools/wxwidgets/wx-config is link to packages instead to lib
> * tools/wxwidgets/wxrc is link to packages instead to lib
> * lib/wx/config/gtk3-unicode-static-3.1  `--cppflags`/`--cflags`/`--cxxflags` options, the output of the include folders are linked to packages instead of include folder.
> 
> Can you fix it?
> 
> image image

@dg0yt Can you help to fix this problem? It vcpkg problem, not upstream problems.

**Pull request merged**